Sources/PrivoSDK/internal/age-gate/AgeGateStorage.swift

```diff
@@ -76,14 +74,59 @@ internal class AgeGateStorage {
         return nil
     }
     
-    func getFpId() async -> String {
-        if let fpId = keychain.get(getFpIdKey()) { return fpId }
-        let fingerprint = DeviceFingerprint()
-        let response = await api.generateFingerprint(fingerprint: fingerprint)
-        //TO DO: need to implement for managing way when there is no id
-        guard let id = response?.id else { return "" }
-        keychain.set(key: getFpIdKey(), value: id)
-        return id
+    // MARK: FingerprintStorage
+    
+    var fpid: String? {
+        get {
+            return keychain.get(getFpIdKey())
+        }
+        set {
+            if let newValue {
+                keychain.set(key: getFpIdKey(), value: newValue)
+            } else {
+                keychain.delete(getFpIdKey())
+            }
+        }
+    }
+}
+
+protocol IFpidStorage {
+    var fpid: String? { get set }
+}
+
+protocol IFpidService {
+    var fpid: String? { get async }
+}
+
+class FpidService: IFpidService {
+    private let source: IRest
+    private var cache: IFpidStorage
+    
+    init(source: IRest = Rest.shared,
+         cache: IFpidStorage = AgeGateStorage()) {
+        self.source = source
+        self.cache = cache
+    }
+    
+    var fpid: String? {
+        get async {
+            return await getFpId()
+        }
     }
     
+    private func getFpId() async -> String? {
+        if let fpid = cache.fpid {
+            return fpid
+        }
+        
+        let rawFpid = DeviceFingerprint()
+        let fpidResponse = await source.generateFingerprint(fingerprint: rawFpid)
+        
+        guard let fpid = fpidResponse?.id else {
+            return nil
```

Comment:
looks like for situations when we don't have FP ID we need to send AnalitycEvent

Reply:
Thought about it and was ready to implement it, but stopped. And this is the reason:
```
    private func getFpId() async -> String? {
        if let fpId = storage.fpId {
            return fpId
        }
        
        let rawFpId = DeviceFingerprint()
        let fpIdResponse = await source.generateFingerprint(fingerprint: rawFpId)
        
        guard let fpId = fpIdResponse?.id else {
            // should sent analytic error here, but...
            return nil
        }
        storage.fpId = fpId

        return fpId
    }
```
We should be sent an analytic error if fpId returns nil. But
```
    func generateFingerprint(fingerprint: DeviceFingerprint) async -> DeviceFingerprintResponse? {
        let url = PrivoInternal.configuration.authBaseUrl.appending(.api).appending(.v1_0).appending(.fingerprint)
        typealias R = DataResponse<DeviceFingerprintResponse,AFError>
        let result: R = await session.request(url, method: .post, parameters: fingerprint, encoder: JSONParameterEncoder.default)
        // ... but if something really happened, try trackPossibleAFError already send the analytic error
        trackPossibleAFError(result.error,  result.debugDescription, result.response?.statusCode)
        return result.value
    }
```

if we get it wrong, trackPossibleAFError already did this work.